Orleans.Consensus/Log/StreamLog.cs

```diff
@@ -0,0 +1,191 @@
+﻿namespace Orleans.Consensus.Log
+{
+    using Orleans.Consensus.Contract.Log;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    public class StreamLog<TOperation> : IPersistentLog<TOperation>, IDisposable
+    {
+        const int CACHE_SIZE = 1000;
+
+        Stream stream;
+        ISerializer<LogEntry<TOperation>> serializer;
+        BTree<long, long> bTreeIndex;
+        IDictionary<long, LogEntry<TOperation>> cache;
+
+        public StreamLog(Stream stream, ISerializer<LogEntry<TOperation>> serializer)
+        {
+            if (null == stream) throw new ArgumentNullException(nameof(stream));
+            if (null == serializer) throw new ArgumentNullException(nameof(serializer));
+
+            this.stream = stream;
+            this.serializer = serializer;
+            this.bTreeIndex = new BTree<long, long>();
+            this.cache = new Dictionary<long, LogEntry<TOperation>>(CACHE_SIZE);
+            BuildIndex();
+        }
+
+        void BuildIndex()
+        {
+            stream.Position = 0;
+            var first = true;
+            while (stream.Position < stream.Length)
+            {
+                var position = stream.Position;
+                var entry = serializer.Deserialize(stream).Result;
+                this.bTreeIndex.Insert(entry.Id.Index, position);
+                AddToCache(entry);
+                this.LastLogEntryId = entry.Id;
+                if (first)
+                {
+                    this.FirstLogEntryId = entry.Id;
+                    first = false;
+                }
+            }
+        }
+
+        void AddToCache(LogEntry<TOperation> entry)
+        {
+            this.cache.Add(entry.Id.Index, entry);
+            while (cache.Count > CACHE_SIZE)
+            {
+                // remove keys with lower indexes
+                cache.Remove(cache.Keys.Min());
```

Comment:
(Not blocking) This seems expensive: it looks like it visits up to 1000 entries for each add. Maybe we should use a different data structure for the cache.